contracts/beacon/Voucher.sol

```diff
@@ -27,7 +33,20 @@ contract Voucher is OwnableUpgradeable, IVoucher {
         uint256 _expiration;
         uint256 _type;
         mapping(address => bool) _authorizedAccounts;
-        mapping(bytes32 dealId => uint256) _sponsoredAmounts;
+        // Save all deals matched by the voucher to be able to
+        // refund requesters for non sponsored tasks.
+        mapping(bytes32 dealId => VoucherMatchedDeal) _voucherMatchedDeals;
+        // Save refunded tasks to disable replay attacks.
+        mapping(bytes32 taskId => bool) _claimedTasks;
+    }
```

Comment:
I propose to move `mapping(bytes32 taskId => bool) _refundedTasks;` to VoucherHub (for example in order to protect against double refund)

Reply:
That way we need to move deal/task checking logic to the VoucherHub.

Reply:
Yes exactly!